client/src/main/kotlin/ClientMain.kt

```diff
@@ -0,0 +1,470 @@
+package client
+
+import common.*
+import io.ktor.network.selector.*
+import io.ktor.network.sockets.*
+import io.ktor.utils.io.*
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.runBlocking
+import kotlinx.serialization.json.Json
+
+fun main(args: Array<String>) {
+    val host = args[0]
+    val port = args[1].toInt()
+    val place = Place.getPlace(args[2].toInt() - 1)
+
+    runBlocking {
+        client(host, port, place)
+    }
+}
+
+private fun cardsToString(cards: Set<Card>): String {
+    return cards.joinToString(", ") { it.toString() }
+}
+
+private fun cardsToString(cards: List<Card>): String {
+    return cards.joinToString(", ") { it.toString() }
+}
+
+private fun nullableCardsToString(cards: List<Card?>): String {
+    return cards.joinToString(", ") { it?.toString() ?: "null" }
+}
+
+suspend fun client(host: String, port: Int, place: Place) {
+    val (input, output, toServer) = connectToServer(host, port, place)
+    val gameState = ClientGameState()
+    while (true) {
+        var serverMsgJson = input.readUTF8Line() ?: throw Exception("No message received")
+        when (val serverMsg = Json.decodeFromString<Message>(serverMsgJson)) {
+            is Message.Rozdanie -> {
+                gameState.myCards = serverMsg.cards.toMutableList()
+                println("Your cards: ${cardsToString(gameState.myCards.toList())}")
+            }
+            else -> {
+                throw Exception("Unexpected message from server, expected cards")
+            }
+        }
+
+        licytacja(gameState, input, output, place)
+
+        val musik = seeMusik(gameState, place, input).toList()
+
+        processMusik(gameState, input, output, place, musik)
+
+        getFinalPoints(gameState, input, output, place)
+
+        gameloop(gameState, input, output, place)
+
+        println("Points: ${gameState.points.joinToString(", ")}, you are $place player")
+
+        if (!gameState.continueGame()) {
+            break
+        }
+    }
+}
+
+
+
+internal suspend fun play1round(lewStarter: Place, gameState: ClientGameState, licytacjaWinnerPoints: Int, myPlace: Place, input: ByteReadChannel, output: ByteWriteChannel): Pair<Place, Int> {
+    // TODO("VALIDATE CARD")
+    var playing = lewStarter
+    val cardsPlayed: Array<Card?> = Array(3) { null }// dummy cards
+    var newLicytacjaWinnerPoints = licytacjaWinnerPoints
+    var firstSuit: CardSuit? = null
+    repeat(3) {
+        println()
+        when(playing) {
```

Comment:
ta funkcja jest bardzo długa i przeplata co najmniej trzy poziomy abstrakcji: logikę gry, logikę prezentacji (wypisywanie i wczytywanie inputu) oraz komunikację z serwerem. Proszę to podzielić w taki sposób, żeby dało się każdą z nich osobno testować (i napisać testy jednostkowe)